File: 0082A.Double.Cola/solution.cpp
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

int main( void )
{
    // peopl's names
    vector< string > names;
    names.push_back( "Sheldon" );
    names.push_back( "Leonard" );
    names.push_back( "Penny" );
    names.push_back( "Rajesh" );
    names.push_back( "Howard" );

    // the nth can
    int n;
    cin >> n;

    int a = 0, s = names.size(), pa = 0;
    for ( int i = 1;; ++i )
    {
        a += s;
        if ( n > a )
        {
            pa = a;
            s += s;
            continue;
        }

        int ind = ( n - pa - 1 ) / static_cast< int >( pow( 2.0, i - 1 ) + 0.5 );

        // output the name of people who drinks the nth can
        cout << names[ ind ] << endl;
        break;
    }

    return 0;
}

File: 0082A.Double.Cola/solution.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

var reader *bufio.Reader = bufio.NewReader(os.Stdin)
var writer *bufio.Writer = bufio.NewWriter(os.Stdout)

func Fprintf(format string, x ...interface{}) {
	fmt.Fprintf(writer, format, x...)
}

func Fscanf(format string, x ...interface{}) {
	fmt.Fscanf(reader, format, x...)
}

func Solution() {
	names := []string{
		"Sheldon",
		"Leonard",
		"Penny",
		"Rajesh",
		"Howard",
	}

	var rank int
	Fscanf("%d\n", &rank)

	accumulation := 0
	nameCount := len(names)
	prevAccumulation := 0
	for power := 1; ; power++ {
		accumulation += nameCount
		if rank > accumulation {
			prevAccumulation = accumulation
			nameCount += nameCount
			continue
		}

		nameIndex := (rank - prevAccumulation - 1) / int(math.Pow(2.0, float64(power-1))+0.5)
		Fprintf("%s\n", names[nameIndex])
		break
	}
}

func main() {
	Solution()
	writer.Flush()
}
